src/vmm/src/devices/legacy/serial.rs

```diff
@@ -264,127 +277,63 @@ impl<W: std::io::Write> MutEventSubscriber
     }
 }
 
-impl<W: Write + Send + 'static> BusDevice
-    for SerialWrapper<EventFdTrigger, SerialEventsWrapper, W>
+impl<I: Read + AsRawFd + Send + Debug + 'static>
+    SerialWrapper<EventFdTrigger, SerialEventsWrapper, I>
 {
-    fn read(&mut self, offset: u64, data: &mut [u8]) {
+    pub fn bread(&mut self, offset: u64, data: &mut [u8]) {
         if data.len() != 1 {
-            self.serial.events().metrics.missed_read_count.inc();
+            METRICS.uart.missed_read_count.inc();
             return;
         }
         data[0] = self.serial.read(offset as u8);
     }
 
-    fn write(&mut self, offset: u64, data: &[u8]) {
+    pub fn bwrite(&mut self, offset: u64, data: &[u8]) {
         if data.len() != 1 {
-            self.serial.events().metrics.missed_write_count.inc();
+            METRICS.uart.missed_write_count.inc();
             return;
         }
         if let Err(err) = self.serial.write(offset as u8, data[0]) {
             // Counter incremented for any handle_write() error.
             error!("Failed the write to serial: {:?}", err);
-            self.serial.events().metrics.error_count.inc();
+            METRICS.uart.error_count.inc();
         }
     }
 }
 
 #[cfg(test)]
 mod tests {
-    use std::io;
-    use std::sync::{Arc, Mutex};
-
     use utils::eventfd::EventFd;
 
     use super::*;
 
-    #[derive(Clone)]
```

Comment:
This test is significantly altered by the changes, so much so that it becomes difficult for it to be useful as it requires so much custom code that it is barely touching non-test code.

Reply:
In this case since the functionality has been reduced so this should not cause a notable reduction in test coverage.

The overall reduction in test coverage in the PR is mostly due to additional lines generated by `Debug` implementations. 1.6% ~= 1000 lines, if we consider each `Debug` implementation generates ~10 lines and there are many `derive(Debug)` in this PR (atleast 60) this accounts for most of the difference. Remaining difference may also be a result of it being close to needing to be updated.